enhancements/kube-apiserver/enforce-labels.md

```diff
@@ -0,0 +1,123 @@
+
+As telco platforms are still evolving, so are the labels requirements. Those are likely to increase in volume. Managing those through the helm chart is not practical for CNF vendors, as they would need to maintain a separate chart for each provider. Moreover, this would only apply to those vendors that deliver their CNF via helm charts.
+
+As an example, some providers expect CaaS specific labels in order to apply the service proxy configurations. These labels are used in the Service Provider’s environment to inject sidecars for service mesh configuration. Other possible uses can be logging or monitoring.
+
+### Goals
+
+- enforce a per project (namespace) labeling scheme
+- ability to configure a cluster-level labeling scheme to enforce on all or a subset of namespaces
+- specify the labeling scheme on resource type and resource level
+- the labeling should complement installation time (eg. Helm or Operator) labeling scheme
+- ensure the labeling scheme is maintained over resource lifecycle (e.g. pod restart)
+- allow customer organizations to update the labeling scheme at any time
+- relabel existing resources upon labeling scheme update
+
+### Non-goals
+
+- the component is only required by specific customers. As such it is NOT to be shipped with every openshift installation, but will be provided as an optional OLM component.
+
+## Proposed solutions
+
+The prefered solution now is to use Gatekeeper:
+https://github.com/open-policy-agent/gatekeeper/
+
+Gatekeeper is the prefered solution. Currently it does not meet all our requirements, but appropriate features are either at design stage, or have been agreed to by Gatekeeper as feasible to add (as opt-in features). An additional bonus is that there are multiple teams interested in having Gatekeeper as part of openshift.
+A custom solution was evaluated, but the idea was discarded.
+
+
+### Gatekeeper
+
+https://github.com/open-policy-agent/gatekeeper/
+
+Gatekeeper is an opensource constraint policy enforcement tool.
+So far gatekeeper only provides a validation webhook, but has advanced plans for a mutation webhook. We're in discussion with gatekeeper people (smythe@google.com) about extending this to provide updates to existing resources in reaction to MutationTemplate changes.
+
+PROs:
+
+- advanced plans to provide a mutation webhook implementing a MutationTemplate (design doc [here|https://docs.google.com/document/d/1G7WgZKx1Y3VOTUjrqn7DjDaZgSKCIZowILm_I6psrw0/edit?pli=1#
+])
+- seem to be interested in adding a updates to existing resources in reaction to MutationTemplate changes (but no hard commitment on this)
+- a regularly release opensource project with  community backed by companies like google, ms, RH, ...
+- multiple Red Hat teams want to have this as part of Openshift
+
+CONs:
+- don't provide any functionality we need yet
+- reactive resource updates still under design
+- alpha version of the mutation webhook only available by the end of the year
```

Comment:
missing a discussion about the risk for OpenShift if these
- don't make it
- are not good enough as alpha
- don't have the features we need.

I heard that our timeline for this topic is also in a similar range of "end of the year".

I would much more comfortable with that if we were the ones doing the upstream implementation. But I understand that we aren't?

Reply:
Added a section called "Risk assesment".
We are actually doing the upstream implementation (although we have no RH maintainer)